Templated SourceHook

Mooshua opened the pull request:
Introduces one massive template for generating SourceHook managers. The goal is for this to simplify the definition of SourceHooks long-term by replacing macros with templates.

**Draft until API is settled upon, but is most likely merge-able now.**

### Introducing SourceHook::Hook
SourceHook::Hook is a template that implements the previous functionality of the non-vafmt `SH_DECL` macros. It uses the following template parameters in it's implementation:
- `ISourceHook** SH` - Pointer to SourceHook (eg `&g_SHPtr`)
- `typename Interface` - The interface to hook
- `Result (Interface::Method)(Args...)` - The argument to hook (eg `&IEngine::DoThing`
- `typename Result` - the return type or `void`
- `typename ...Args` - all arguments, specified as a varardic type pack.

Example usage:
```c++
//                                           Interface       Method                     Result
//                                           vvvvv           vvvvv                      vvvv
auto OnGameInit = SourceHook::Hook<&g_SHPtr, IServerGameDLL, &IServerGameDLL::GameInit, bool>::Make();
auto OnLevelInit = SourceHook::Hook<&g_SHPtr, IServerGameDLL, &IServerGameDLL::LevelInit, bool, const char*, const char*, const char*, const char*, bool, bool>::Make();

void AddHooks()
{
	OnGameInit->Add(g_PLID, server, false, SH_MEMBER(this, &SourceProvider::Hook_GameInit));
	OnLevelInit->Add(g_PLID, server, false, SH_MEMBER(this, &SourceProvider::Hook_LevelInit));
}
```

GAMMACASE commented:
Very neat stuff, but would throw my 2 cents here, what do you think on renaming exposed struct ``Hook`` to something more verbose, like ``SH_Hook``? Just to prevent any potential name collisions and further confusion due to it on the user end.

Mooshua commented:
> Very neat stuff, but would throw my 2 cents here, what do you think on renaming exposed struct `Hook` to something more verbose, like `SH_Hook`? Just to prevent any potential name collisions and further confusion due to it on the user end.

How about "HookFactory" or "SourceHookFactory", since that's effectively what it is?

dvander commented:
Thanks for doing this! Will take a look this weekend hopefully.

Namespaces are good avoiding for name collisions.

Mooshua commented:
Added varardic hook macros, and I decided to just put the helpers into the `SourceHook` namespace. Hopefully this doesn't cause issues with plugins.

Split the original `Hook` template into four new templates: One for SourceHook -> Delegate APIs (`HookHandlerImpl`), one for Plugin -> SourceHook APIs (`HookCoreImpl`, and two handling the ABI for the raw hooks (`HookImpl` and `FmtHookImpl`). They're still a little rough around the edges and need some boundary refinement.

As a result of this, varfmt hooks are now supported, with similar handling to macro varfmt hooks:
```c++
auto OnClientCommand = SourceHook::FmtHook<IVEngineServer, &IVEngineServer::ClientCommand, void, edict_t*>::Make();

void SamplePlugin::Hook_SendClientCommand(edict_t* pEntity, const char* pszCommand)
{
    // pszCommand = printf(fmt, ...)
}
```

Still thinking about how I want to tackle manual hooks; open to suggestions.

dvander commented:
Not everything has to be solved in one PR! Happy to take incremental improvements. API changes don't get frozen until release.

Kenzzer commented on core/sourcehook/generate/sourcehook.hxx:
```diff
@@ -566,6 +570,730 @@ namespace SourceHook
+				virtual~CMyDelegateImpl() {}
+				Result Call(Args... args) { return _delegate(args...); }
+				void DeleteThis() { delete this; }
+				bool IsEqual(ISHDelegate *pOtherDeleg) { return _delegate == static_cast<CMyDelegateImpl *>(pOtherDeleg)->_delegate; }
```
This looks very suspicious
`static_cast<CMyDelegateImpl *>`
The other delegate is not guaranteed to be a `CMyDelegateImpl`, this might not be the case here and now. But from an API standpoint `IsEqual` can be called with any classes implementing `SHDelegate`, one notable such class could be dynamically generated hooks.

I suggest swapping this out for a `dynamic_cast` or return false all the time.

Mooshua replied:
That’s actually what the [macros do](https://github.com/alliedmodders/metamod-source/blob/8abbfbb8106bfe79b0096d981bbc337e58a44052/core/sourcehook/generate/sourcehook.hxx#L1780).

Would there be any behavior change using dynamic_cast if any plugins running under Metamod were compiled without RTTI?

Kenzzer replied:
Relaying my thoughts from discord.

After carefully reading SH, `IsEqual` is currently only invoked when a hook is removed. Meaning whatever delegate is passed will be a `CMyDelegate`. So you can actually ensure the class type the function will be invoked with... Until we start accounting dynamic hooks, their `ISHDelegate` implem can be empty, and if those are ever removed. Then `ISHDelegate` will not be a `CMyDelegate`.

Ultimately since `_delegate` is only `+4` or `+8` from the this ptr, IF `pOtherDeleg` isn't a `CMyDelegate` it won't fall too far away from the address and most likely still be in a read allowed space. Buuuut there's still a risk, and it might mislead future contributors about `bool IsEqual(ISHDelegate *pOtherDeleg)` like it did for you.

Then there's the issue you raise with RTTI and dynamic cast. And this currently seems like an unsolvable problem. Not without either modifying the virtual function definition to provide more details about the delegate, or make a new one.

dvander replied:
IsEqual() should only be called on ISHDelegates within the same plugin (eg binary) *and* the same HookManPubFunc, which means there's no way it'll be called on ISHDelegates that were macro-expanded/compiled in different places.

We have to wrap our heads around this every time it comes up, so it's probably worth a comment in the code somewhere, and maybe a test mixing hook types (eg a macro hook and a template hook on the same vtable entry) proving that there are no cross-calls.

Kenzzer replied:
Oh yeah true, I hadn't realised `HookManPubFunc` guards against that issues.

As for the test I can do that in the revised sourcehook test suite I've added in my own PR.

Mooshua replied:
Cool! I'll mark this as resolved then. Sticking with static casts for now

Kenzzer commented on core/sourcehook/generate/sourcehook.hxx:
```diff
@@ -566,6 +570,730 @@ namespace SourceHook
+			template<ISourceHook** SH, typename Invoker, typename InstType>
+			static Result HookImplCore(InstType* Instance, void* self, Args... args)
+			{
+				//	Note to all ye who enter here:
+				//	Do not, do NOT--DO NOT: touch "this" or any of our member variables.
+				//	Hands off bucko. USE THE STATIC INSTANCE! (thisptr is undefined!!)
```
That comment should be unnecessary, given that the function is static, `this` keyword is out of bounds.

Mooshua replied:
Ah yes that was inherited from when that method used to be virtual. Will smite that, thanks

Kenzzer commented on core/sourcehook/generate/sourcehook.hxx:
```diff
@@ -566,6 +570,730 @@ namespace SourceHook
+				ResultType override_ret;
+				ResultType current_ret;
+
+				IMyDelegate *iter;
```
This seems to be remnant of the macros ? But using `::SourceHook::ISHDelegate` here should be fine. That will avoid the three `static_cast` down below. A nitpick mostly, disregard if you don't want to fix it.

Mooshua replied:
Straight from the macro, baby! Thanks

Kenzzer reviewed (changes requested):
First thing first, that's an awesome PR everything is thoroughly documented. And SourceHook really needed a makeover with templates, this is amazing.

Also since I happen to also be modifying SourceHook in another PR. I think I'm familiar enough with the framework to give a tiny review. There's one problem spotted, everything else is nitpicks, I might do another review if I've time to try this out in a project.

Mooshua commented:
Don’t merge just yet—-just found some issues with reference return handling. Should be a quick fix.

Mooshua commented on core/sourcehook/generate/sourcehook.hxx:
```diff
@@ -566,6 +570,737 @@ namespace SourceHook
+		*	@param iface the interface pointer to hook
+		*	@param post true when post-hooking, false when pre-hooking.
+		*	@param handler the handler that will be called in place of the original method
+		*	@param mode the hook mode - Hook_Normal will only hook this instance, while others alter the global virtual table.
```
Thank you!

dvander reviewed (commented):
This is absolutely fantastic! I'm impressed with how straightforward it is and how much nicer the actual interface is. Best of all it does not touch the ABI as far as I can tell.

Aside from some minor nits, I'd be happy to take this on 2.0 and/or 1.12 (though you probably want to bake it for a little bit before doing a backport).

I think one thing is needed before committing to this on 2.0: some tests. They don't have to be that advanced, but some kind of addition to the SourceHook test suite would be great. If for no other reason to document the new API a little.

And to commit on 1.12: Take a new MM:S and old SourceMod and make sure they run together, as a smoke test that the ABI hasn't changed.

Mooshua commented on core/sourcehook/sourcehook.h:
```diff
@@ -566,6 +570,737 @@ namespace SourceHook
+
+			static Result Dereference(const RefSafeResult* arg)
+			{
+				Result res = *arg;
```
You would think, but this actually breaks compatibility with reference returns somehow.

dvander replied:
Mysterious... add a comment then?

Mooshua replied:
Nevermind??? Must have been something else broken, GCC handles it fine now.

dvander commented:
There's also a build error:

```
/home/runner/work/metamod-source/metamod-source/metamod-source/core/sourcehook/sourcehook.h:1279:5: error: use of undeclared identifier 'vsnprintf'
                                vsnprintf(buf, sizeof(buf), fmt, ap);
```

I'm not sure why this didn't get pulled in as part of `<cstdarg>`.

Mooshua commented on core/sourcehook/sourcehook.h:
```diff
@@ -566,6 +570,737 @@ namespace SourceHook
+
+	//	You're probably wondering what the hell this does.
+	//	https://stackoverflow.com/questions/11709859/how-to-have-static-data-members-in-a-header-only-library/11711082#11711082
+	//	I hate C++.
```
I think this was back when I was trying to get this to work on C++11, before I realized that was plain impossible. I'll look into an `inline static`.

Mooshua replied:
Going to keep this as a hack for the time being just because I don't understand the semantic differences between this and `inline static`. Would be more than willing to do either I just dont know what is and isn't UD these days :/

Mooshua commented on core/sourcehook/sourcehook.h:
```diff
@@ -566,6 +570,737 @@ namespace SourceHook
+	 * @tparam Args All arguments passed to the original 
+	 */
+	template<ISourceHook** SH, Plugin* PL, typename Interface, auto Method, typename Result, typename... Args>
+	struct HookImpl : HookCoreImpl<
```
Yes, it is inherited in the PLUGIN_GLOBALVARS() macro to hide the `ISourceHook**` and `Plugin*` template arguments.

Mooshua commented on core/ISmmPlugin.h:
```diff
@@ -475,12 +475,24 @@ using namespace SourceMM;
+	namespace SourceHook                                                                                         \
+	{                                                                                                            \
+		template <typename Interface, auto Method, typename Result, typename... Args>                            \
+		struct Hook : public ::SourceHook::HookImpl<&g_SHPtr, &g_PLID, Interface, Method, Result, Args...>       \
```
Done! Great idea, final has been added to FmtHookImpl and HookImpl

Mooshua commented on core/provider/source/provider_source.cpp:
```diff
@@ -44,14 +44,14 @@ static SourceProvider g_SourceProvider;
-SH_DECL_HOOK0(IServerGameDLL, GameInit, SH_NOATTRIB, 0, bool);
-SH_DECL_HOOK6(IServerGameDLL, LevelInit, SH_NOATTRIB, 0, bool, const char*, const char*, const char*, const char*, bool, bool);
-SH_DECL_HOOK0_void(IServerGameDLL, LevelShutdown, SH_NOATTRIB, 0);
+auto OnGameInit = SourceHook::Hook<IServerGameDLL, &IServerGameDLL::GameInit, bool>::Make();
```
I'll keep it as is in the core but I'll probably do it in the sample plugins.

MSWS commented:
LGTM!
